pkg/3scale/amp/operator/option_provider_utils.go

```diff
@@ -0,0 +1,61 @@
+package operator
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type SecretSource struct {
+	client    client.Client
+	namespace string
+}
+
+func NewSecretSource(client client.Client, namespace string) *SecretSource {
+	return &SecretSource{
+		client:    client,
+		namespace: namespace,
+	}
+	// TODO implement caching??
+}
+
+func (s *SecretSource) FieldValue(secretName, fieldName string, def string) (*string, error) {
+	return s.fieldReader(secretName, fieldName, false, false, def)
+}
+
+func (s *SecretSource) FieldValueFromRequiredSecret(secretName, fieldName string, def string) (*string, error) {
+	return s.fieldReader(secretName, fieldName, true, false, def)
+}
+
+func (s *SecretSource) RequiredFieldValue(secretName, fieldName string) (*string, error) {
+	return s.fieldReader(secretName, fieldName, false, true, "")
+}
+
+func (s *SecretSource) RequiredFieldValueFromRequiredSecret(secretName, fieldName string) (*string, error) {
+	return s.fieldReader(secretName, fieldName, true, true, "")
+}
+
+func (s *SecretSource) fieldReader(secretName, fieldName string, secretRequired, fieldRequired bool, def string) (*string, error) {
```

Comment:
We take advantage in the fieldReader method of the fact that a non-existing secret behaves as an existing one.
Thinking about it, with how setSecretBasedOptions is implemented right now, it could happen that the secret might not exist at all and we would not get any errors when providing the options. Could this behavior cause us any problem?

Reply:
For the backend options provider specific case, the `setSecretBasedOptions` does not require the existence of the secret nor the existence of any of the fields, so using `FieldValue` is appropriate. Other option providers might have different requirements and should be using `FieldValueFromRequiredSecret`, `RequiredFieldValue` or `RequiredFieldValueFromRequiredSecret` methods